object_store/src/client/retry.rs

```diff
@@ -190,70 +190,87 @@ impl RetryExt for reqwest::RequestBuilder {
-                                status: StatusCode::NOT_MODIFIED,
-                            })
-                        }
-                        Ok(r) => {
-                            let is_bare_redirect = r.status().is_redirection() && !r.headers().contains_key(LOCATION);
-                            return match is_bare_redirect {
-                                true => Err(Error::BareRedirect),
-                                // Not actually sure if this is reachable, but here for completeness
-                                false => Err(Error::Client {
+                    Ok(r) => {
+                        info!("NEW: Response status: {}", r.status());
+                        info!("NEW: Headers: {:#?}", r.headers());
+                        info!("NEW: Final URL of response: {}", r.url());
+    
+                        match r.error_for_status_ref() {
+                            Ok(_) if r.status().is_success() => {
+                                info!("NEW: Value of is_success(): {}", r.status().is_success());
+                                return Ok(r);
+                            }
+                            Ok(r) if r.status() == StatusCode::NOT_MODIFIED => {
+                                info!("NEW: Entered the StatusCode::NOT_MODIFIED branch, no body");
+                                return Err(Error::Client {
                                     body: None,
-                                    status: r.status(),
+                                    status: StatusCode::NOT_MODIFIED,
                                 })
                             }
-                        }
-                        Err(e) => {
-                            let status = r.status();
-                            if retries == max_retries
-                                || now.elapsed() > retry_timeout
-                                || !status.is_server_error() {
-
-                                return Err(match status.is_client_error() {
-                                    true => match r.text().await {
+                            Ok(r) => {
+                                let is_bare_redirect =
+                                    r.status().is_redirection() && !r.headers().contains_key(LOCATION);
+                                info!("NEW: is_bare_redirect: {} and location: {}", is_bare_redirect, LOCATION);
+                                return match is_bare_redirect {
+                                    true => Err(Error::BareRedirect),
+                                    false => Err(Error::Client {
+                                        body: None,
+                                        status: r.status(),
+                                    }),
+                                };
+                            }
+                            Err(e) => {
+                                // Log the error message
+                                error!("Error in response: {}", e);
```

Comment:
For ease of reading, do you think we should add both `e` and `r` in the log message?